packages/git/src/browser/diff/git-diff-service.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { GitLogOptions } from './git-diff-model';
+import { injectable, inject } from "inversify";
+import { GitRepositoryProvider } from '../git-repository-provider';
+import { Git, Repository, GitFileChange, GitUtils } from '../../common';
+import { FileUri } from '@theia/core/lib/node/file-uri';
+import * as Path from 'path';
+import URI from '@theia/core/lib/common/uri';
+
+@injectable()
+export class GitDiffService {
+
+    constructor(
+        @inject(Git) protected readonly git: Git,
+        @inject(GitRepositoryProvider) protected readonly repositoryProvider: GitRepositoryProvider
+    ) { }
+
+    async getDiff(repository: Repository, options: GitLogOptions): Promise<GitFileChange[]> {
+        const range = this.getRangeArg(options);
+        const args = ['diff', '--name-status', range];
+        if (options.fileUri) {
+            args.push(...['--', GitUtils.getRepositoryRelativePath(repository, new URI(options.fileUri))]);
+        }
+        const changes: GitFileChange[] = [];
+        (await this.git.exec(repository, args)).stdout.split('\0').map(line => line.match(/\S+/g) || []).forEach(fragments => {
+            for (let i = 0; i < fragments.length; i = i + 2) {
+                const status = GitUtils.mapStatus(fragments[i]);
```

Comment:
The `fragments` is not strictly `[kind, file, kind, file, ...]`. But in case it is a move/rename there are two files followed. It creates bogus file entries in the view.
See the screenshot:

<img width="363" alt="screen shot 2018-01-23 at 10 56 30" src="https://user-images.githubusercontent.com/372735/35269449-54f5041e-002c-11e8-8ed4-9210036b7856.png">

Reply:
I will take care of this. I would propose going with `git log` instead of `git diff` but I need to dig into this.